File: frontend/src/api/knowledge.ts
import apiFetch from './client'

export interface KnowledgeBase {
  id: string
  name: string
  description: string | null
  icon: string
  document_count: number
  chunk_count: number
  status: string
  created_by: string
  created_at: string | null
  updated_at: string | null
}

export interface KnowledgeDocument {
  id: string
  knowledge_base_id: string
  title: string
  content: string | null
  content_type: string
  file_name: string | null
  file_size: number | null
  chunk_count: number
  status: string
  created_by: string | null
  created_at: string | null
}

export interface SearchResult {
  chunk_id: string
  content: string
  chunk_index: number
  token_count: number
  document_id: string
  score: number
}

export async function listKnowledgeBases(page = 1, pageSize = 10, search?: string): Promise<{ items: KnowledgeBase[]; total: number }> {
  const qs = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  if (search) qs.set('search', search)
  const resp = await apiFetch(`/api/v1/knowledge-bases?${qs}`, { method: 'GET' })
  return resp.data
}

export async function createKnowledgeBase(name: string, description?: string, icon?: string): Promise<KnowledgeBase> {
  const resp = await apiFetch('/api/v1/knowledge-bases', { method: 'POST', data: { name, description, icon } })
  return resp.data
}

export async function getKnowledgeBase(id: string): Promise<KnowledgeBase> {
  const resp = await apiFetch(`/api/v1/knowledge-bases/${id}`, { method: 'GET' })
  return resp.data
}

export async function getDocuments(kbId: string, page = 1, pageSize = 20): Promise<{ items: KnowledgeDocument[]; total: number }> {
  const resp = await apiFetch(`/api/v1/knowledge-bases/${kbId}/documents?page=${page}&page_size=${pageSize}`, { method: 'GET' })
  return resp.data
}

export async function addDocument(kbId: string, title: string, content: string, contentType = 'text'): Promise<KnowledgeDocument> {
  const resp = await apiFetch(`/api/v1/knowledge-bases/${kbId}/documents`, { method: 'POST', data: { title, content, content_type: contentType } })
  return resp.data
}

export async function deleteDocument(kbId: string, docId: string): Promise<any> {
  const resp = await apiFetch(`/api/v1/knowledge-bases/${kbId}/documents/${docId}`, { method: 'DELETE' })
  return resp.data
}

export async function searchKnowledge(kbId: string, query: string, topK = 5): Promise<{ results: SearchResult[]; query: string; count: number }> {
  const resp = await apiFetch(`/api/v1/knowledge-bases/${kbId}/search?query=${encodeURIComponent(query)}&top_k=${topK}`, { method: 'GET' })
  return resp.data
}

File: backend/backup/structure_optimization/app/api/v1/knowledge.py
"""
知识库 API - CRUD/文档管理/检索
"""
from fastapi import APIRouter, Depends, HTTPException, Query, Body
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models.user import User
from app.services.auth_service import get_current_user
from app.services import knowledge_service

router = APIRouter(tags=["知识库"])


@router.get("/knowledge-bases")
async def list_knowledge_bases(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=50),
    search: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    items, total = await knowledge_service.list_knowledge_bases(db, page=page, page_size=page_size, search=search)
    return {
        "items": [_format_kb(k) for k in items],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.post("/knowledge-bases")
async def create_knowledge_base(
    name: str = Body(...),
    description: str | None = None,
    icon: str = "📚",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    kb = await knowledge_service.create_knowledge_base(db, name=name, description=description, icon=icon, created_by=current_user.id)
    return _format_kb(kb)


@router.get("/knowledge-bases/{kb_id}")
async def get_knowledge_base(
    kb_id: str,
    db: AsyncSession = Depends(get_db),
):
    kb = await db.get(knowledge_service.KnowledgeBase, kb_id)
    if not kb:
        raise HTTPException(status_code=404, detail="知识库不存在")
    return _format_kb(kb)


@router.get("/knowledge-bases/{kb_id}/documents")
async def get_documents(
    kb_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    items, total = await knowledge_service.get_knowledge_base_documents(db, kb_id, page=page, page_size=page_size)
    return {
        "items": [_format_doc(d) for d in items],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.post("/knowledge-bases/{kb_id}/documents")
async def add_document(
    kb_id: str,
    title: str = Body(...),
    content: str = Body(...),
    content_type: str = "text",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        doc = await knowledge_service.add_document(
            db, kb_id, title=title, content=content, content_type=content_type, created_by=current_user.id,
        )
        return _format_doc(doc)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/knowledge-bases/{kb_id}/documents/{doc_id}")
async def delete_document(
    kb_id: str,
    doc_id: str,
    db: AsyncSession = Depends(get_db),
):
    success = await knowledge_service.delete_document(db, doc_id)
    if not success:
        raise HTTPException(status_code=404, detail="文档不存在")
    return {"message": "文档已删除"}


@router.get("/knowledge-bases/{kb_id}/search")
async def search_knowledge(
    kb_id: str,
    query: str = Query(..., min_length=1),
    top_k: int = Query(5, ge=1, le=20),
    db: AsyncSession = Depends(get_db),
):
    results = await knowledge_service.search_knowledge(db, kb_id, query=query, top_k=top_k)
    return {"results": results, "query": query, "count": len(results)}


def _format_kb(k):
    return {
        "id": k.id,
        "name": k.name,
        "description": k.description,
        "icon": k.icon,
        "document_count": k.document_count,
        "chunk_count": k.chunk_count,
        "status": k.status,
        "created_by": k.created_by,
        "created_at": k.created_at.isoformat() if k.created_at else None,
        "updated_at": k.updated_at.isoformat() if k.updated_at else None,
    }


def _format_doc(d):
    return {
        "id": d.id,
        "knowledge_base_id": d.knowledge_base_id,
        "title": d.title,
        "content": d.content[:200] + "..." if d.content and len(d.content) > 200 else d.content,
        "content_type": d.content_type,
        "file_name": d.file_name,
        "file_size": d.file_size,
        "chunk_count": d.chunk_count,
        "status": d.status,
        "created_by": d.created_by,
        "created_at": d.created_at.isoformat() if d.created_at else None,
    }
